chore: add incremental type checking with mypy

denismaciel opened the pull request:
The PR sets up mypy to be adopted incrementally in the code base.

The idea is: we make mypy ignore all files. We then annotate one file at a time. As we annotate files, we remove them from being ignored by mypy.

I have type-annotated the `apy.config` module while ignoring all other modules.

The workflow for annotating an extra file is:

1. Remove the file from the ignore list in `pyproject.toml`
2. Run `mypy src` to identify the violations
3. Fix the violations
4. Be happy and type-safe

Do you think this works for you, @lervag?

If you're happy with the approach, I will annotate/fix other files and add a type-check step to the Github Action.

Looking forward to your feedback.

Greetings from Hamburg!

denismaciel commented on src/apy/config.py:
```diff
@@ -2,15 +2,16 @@
 
-    if path := os.environ.get("ANKI_BASE"):
-        return path
+    if path_as_str := os.environ.get("ANKI_BASE"):
```
mypy complains here because in the code below `path` changes its type from `str` to `pathlib.Path`.

lervag replied:
It's unclear what you mean. Will mypy complain after these changes or was this particular change necessary _because_ mypy complained?

denismaciel replied:
mypy was complaining about the previous version of the code. 

All the changes I did in the PR were solely to please mypy. No change in behavior.

Just to clarify what the problem was in the previous version of the code:

* the variable `path` starts off as a string
* down the function, it becomes a `pathlib.Path`
* mypy doesn't allow a variable to change its type and complained

lervag commented on pyproject.toml:
```diff
@@ -48,3 +51,19 @@ disable = [
+
+[tool.mypy]
+exclude = ".venv"
+strict = true
```
Could you explain these settings for me (or point me to a good reference)?

denismaciel replied:
Here's the docs for [`--strict`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-strict).

lervag commented on pyproject.toml:
```diff
@@ -48,3 +51,19 @@ disable = [
+module = [
+    "apy.anki",
+    "apy.cli",
+    # "apy.config",
```
Perhaps we could add a comment above line 59 to explain wht the point of these overrides are? And I guess we can _delete_ the finished lines instead of commenting them out?

lervag commented on src/apy/config.py:
```diff
@@ -34,21 +35,19 @@ def get_base_path():
+    "presets": {},
+    "profile_name": None,
+    "query": "tag:marked OR -flag:0",
+}
```
Was this change necessary, or only useful to indicate the point of the dict? Also, I think I'd prefer an empty line after line 48 here.

denismaciel replied:
mypy was complaining because it needed an explicit annotation for `default` in the line below:

```
for required, default in ...
```

I guess I could have just solved it by adding one line:

```
default: Any
```

but I feel using a dictionary instead of a tuple conveys better the idea of it being a default configuration.

Wdyt?

(I will add a new line)

lervag replied:
I think this looks good; thanks!

lervag reviewed (commented):
> The PR sets up mypy to be adopted incrementally in the code base.

Great, thanks!

> The idea is: we make mypy ignore all files. We then annotate one file at a time. As we annotate files, we remove them from being ignored by mypy.

Good idea.

> The workflow for annotating an extra file is: …

This sounds reasonable as a workflow. I guess it shouldn't take that long either, it is not really so much code.

> Do you think this works for you, @lervag?

Yes!

> If you're happy with the approach, I will annotate/fix other files and add a type-check step to the Github Action.

Good; I guess adding the type-check step to the GHA could be useful as part of this PR?

> Greetings from Hamburg!

Greetings from Norway :)

denismaciel commented:
@lervag I have addressed your comments and added mypy to CI.

I was also thinking we could make the `cfg` dictionary a `dataclasses.dataclass` eventually.

lervag commented on .github/workflows/ci.yaml:
```diff
@@ -29,6 +29,8 @@ jobs:
       - name: Run black
         run: poetry run black --check .
+      - name: Run mypy
+        run: poetry run mypy src
```
How about type checking the tests?

denismaciel replied:
From experience, I haven't gotten too much value of type checking tests.

I have nothing against it and cannot make a strong argument against it either.

I'd suggest to restrict this PR to type checking src and you could play around with typing the checks to see if it's something you'd like.

ckp95 replied:
the idea behind static type checking is to enforce certain invariants across all possible program states. function foo always returns a list of strings, things like that. unit tests are for checking behaviour on specific inputs. they're complementary.

you don't need to typecheck your tests because they're already only running through a specifically enumerated set of inputs: the test cases. so you don't need that level of generality. furthermore, tests frequently benefit from dynamic features like monkeypatching, so they don't mesh well with type checkers.

so I agree that there's little sense in running mypy on the tests directory.

lervag replied:
Thanks for the info, @ckp95! I appreciate it, and I agree it makes sense to keep the typechecking only on the main source code.

lervag reviewed (commented):
> I was also thinking we could make the cfg dictionary a dataclasses.dataclass eventually.

Seems like a good idea!

I think this is very good! The final thing I'm thinking about is if we should include the tests here?

ckp95 commented:
I was actually planning to make a PR with a refactor of the config to use a dataclass. It's cleaner because you see the fields and the types all in one place. type checker can work with it better than a heterogeneously typed dict.

ckp95 commented:
the idea is something like this

```python
import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional, Any

from platformdirs import user_config_path, user_data_path


def env_path(name: str, must_exist_if_specified=True) -> Optional[Path]:
    if (raw_path := os.environ.get(name)) is None:
        return None

    if must_exist_if_specified and not (path := Path(raw_path)).exists():
        raise FileNotFoundError(f"No such path: {name}={path}")

    return path


def get_raw_config() -> dict[str, Any]:
    cfg_dir = env_path("APY_CONFIG") or user_config_path("apy")
    if not (cfg_path := cfg_dir / "apy.json").exists():
        return {}

    return json.loads(cfg_path.read_text())


def get_base_path() -> Path:
    return env_path("APY_BASE") or env_path("ANKI_BASE") or user_data_path("Anki2")


@dataclass
class Config:
    base_path: Path = field(default_factory=get_base_path)
    img_viewers: dict[str, list[str]] = field(
        default_factory=lambda: {"svg": ["display", "-density", "300"]}
    )
    img_viewers_default: list[str] = field(default_factory=lambda: ["feh"])
    markdown_models: list[str] = field(default_factory=lambda: ["Basic"])
    presets: dict[str, dict] = field(default_factory=dict)
    profile: Optional[str] = None
    query: str = "tag:marked OR -flag:0"

    def __post_init__(self) -> None:
        if not self.base_path.exists():
            raise FileNotFoundError(f"Could not find Anki base path: {self.base_path}")

        self.base_path = self.base_path.absolute()
        self.presets.setdefault("default", {"model": "Basic", "tags": []})


cfg = Config(**get_raw_config())
```

haven't tested this or modified the other modules to account for the changes yet.

denismaciel commented:
From experience, I haven't gotten too much value of type checking tests. 

I have nothing against it and cannot make a strong argument against it either.

I'd suggest to restrict this PR to type checking `src` and you could play around with typing the checks to see if it's something you'd like.

ckp95 commented:
Would it make more sense for me to push this dataclass config stuff to this PR, or to make a separate one?

lervag commented:
Could you make a new PR? I'll merge this one probably later today. If you make the PR based on this branch and later rebase it on master after this is merged, then it should be quite clean.
